src/osp/Active/SysVehicle.cpp

```diff
@@ -227,20 +219,68 @@ int SysVehicle::deactivate_sat(ActiveScene &scene, SysAreaAssociate &area,
     return 0;
 }
 
+void SysVehicle::add_machines_to_object(ActiveEnt partEnt, ActiveEnt objEnt,
+    std::vector<PrototypeMachine> const& protoMachines,
+    std::vector<BlueprintMachine> const& blueprintMachines,
+    std::vector<unsigned> const& machineIndices)
+{
+    if (machineIndices.empty()) { return; }
+    auto &compMachines =
+        m_scene.get_registry().get_or_emplace<ACompMachines>(partEnt);
+
+    for (unsigned i : machineIndices)
+    {
+        BlueprintMachine const& bpMachine = blueprintMachines[i];
+        PrototypeMachine const& protoMachine = protoMachines[i];
+
+        MapSysMachine_t::iterator sysMachine
+            = m_scene.system_machine_find(protoMachine.m_type);
 
-ActiveEnt SysVehicle::part_instantiate(PrototypePart& part,
-                                       ActiveEnt rootParent)
+        if (!(m_scene.system_machine_it_valid(sysMachine)))
+        {
+            std::cout << "Machine type: " << protoMachine.m_type << " Not found\n";
+            continue;
+        }
+
+        // Instantiate the machine for the object
+        Machine& machine = sysMachine->second->instantiate(objEnt,
+            protoMachine, bpMachine);
+
+        // Add a PartMachine definition to the root part's ACompMachines
+        compMachines.m_machines.emplace_back(objEnt, sysMachine);
+    }
+}
+
+void SysVehicle::part_instantiate_machines(ActiveEnt partEnt,
+    std::vector<MachineDef> machineMapping,
+    PrototypePart& part, BlueprintPart& partBP)
 {
+    std::vector<PrototypeMachine> const& protoMachines = part.get_machines();
+    std::vector<BlueprintMachine> const& bpMachines = partBP.m_machines;
+    for (auto const& obj : machineMapping)
+    {
+        add_machines_to_object(partEnt, obj.m_machineOwner,
+            protoMachines, bpMachines, obj.m_machineIndices);
+    }
+}
 
+std::pair<ActiveEnt, std::vector<SysVehicle::MachineDef>> SysVehicle::part_instantiate(
+    PrototypePart& part, BlueprintPart& blueprint, ActiveEnt rootParent)
+{
     std::vector<PrototypeObject> const& prototypes = part.get_objects();
     std::vector<ActiveEnt> newEntities(prototypes.size());
 
+    /* A list of MachineDefs, which catalog all part machines and the ActiveEnts
+     * that are created to represent the objects that own them
+     */
+    std::vector<MachineDef> machineMapping;
```

Comment:
machineMapping.reserve(prototypes.size());

Reply:
Another reserve is needed inside the loop as well.

 std::vector<DependRes<Texture2D>> textureResources;
            for (unsigned i = 0; i < drawable.m_textures.size(); i++)
            {